nwb_conversion_tools/datainterfaces/icephys/baseicephysinterface.py

```diff
@@ -0,0 +1,145 @@
+"""Author: Luiz Tauffer."""
+from abc import ABC
+from typing import Optional
+
+from pynwb import NWBFile
+from pynwb.device import Device
+from pynwb.icephys import IntracellularElectrode
+
+from ...basedatainterface import BaseDataInterface
+from ...tools.neo import (
+    get_number_of_electrodes,
+    get_number_of_segments,
+    write_neo_to_nwb,
+)
+from ...utils import (
+    OptionalFilePathType,
+    get_schema_from_hdmf_class,
+    get_schema_from_method_signature,
+    get_base_schema,
+)
+
+
+try:
+    from ndx_dandi_icephys import DandiIcephysMetadata
+
+    HAVE_NDX_DANDI_ICEPHYS = True
+except ImportError:
+    HAVE_NDX_DANDI_ICEPHYS = False
+
+
+class BaseIcephysInterface(BaseDataInterface, ABC):
+    """Primary class for all intracellular NeoInterfaces."""
+
+    neo_class = None
+
+    @classmethod
+    def get_source_schema(cls):
+        source_schema = get_schema_from_method_signature(class_method=cls.__init__, exclude=[])
+        return source_schema
+
+    def __init__(self, file_paths: list):
+        super().__init__(file_paths=file_paths)
+
+        self.readers_list = list()
+        for f in file_paths:
+            self.readers_list.append(self.neo_class(filename=f))
+
+        self.subset_channels = None
+        self.n_segments = get_number_of_segments(neo_reader=self.readers_list[0], block=0)
+        self.n_channels = get_number_of_electrodes(neo_reader=self.readers_list[0])
+
+    def get_metadata_schema(self):
+        metadata_schema = super().get_metadata_schema()
+
+        metadata_schema["properties"]["Icephys"] = get_base_schema(tag="Icephys")
+        metadata_schema["properties"]["Icephys"]["required"] = ["Device", "IntracellularElectrode"]
+        metadata_schema["properties"]["Icephys"]["properties"] = dict(
+            Device=dict(type="array", minItems=1, items={"$ref": "#/properties/Icephys/properties/definitions/Device"}),
+            IntracellularElectrode=dict(
+                type="array",
+                minItems=1,
+                items={"$ref": "#/properties/Icephys/properties/definitions/IntracellularElectrode"},
+            ),
+        )
+        metadata_schema["properties"]["Icephys"]["properties"]["definitions"] = dict(
+            Device=get_schema_from_hdmf_class(Device),
+            IntracellularElectrode=get_schema_from_hdmf_class(IntracellularElectrode),
+        )
+        return metadata_schema
```

Comment:
Wherever possible, we've been trying to move more and more towards the approach seen in the YAML conversion specification feature
 
here: https://github.com/catalystneuro/nwb-conversion-tools/blob/main/nwb_conversion_tools/tools/yaml_conversion_specification/yaml_conversion_specification.py#L44-L51

and here: https://github.com/catalystneuro/nwb-conversion-tools/blob/main/nwb_conversion_tools/schemas/yaml_conversion_specification_schema.json

where all hard-coded schema related things are put in `.json` files instead of written out in code. This really helps for communicating and managing the expected structure as well as visual interpolation for debugging. It also reduces the raw code base.

Note you can still leverage the `get_schema_from_hdmf_class` function yourself to get the text form that would go in such a file.

While we're still in the process of refactoring this across the project, it would be best to get ahead of it on new feature additions such as this interface.

Would you feel up to doing this in this PR, or would you prefer to do it soon in a follow-up?